File: CS502018/pset6/sentimental/credit/credit.py
#import the cs50 library
import cs50

def main():
    #gets the credit card number
    number = cs50.get_string("Number: ")

    #gets the credit card length
    length = len(number)

    #checks if the length of the number is one of the card type
    if length != 15 and length != 16 and length != 13:

        print("INVALID")

        return 1

    #finds the credit card type
    type = credit_type(number)

    #end program is credit_type returned false
    if type == False:
        print("INVALID")
        return 1

    #checks if the card number is valid
    valid = check_sum(number, length)

    #prints the type if card is valid
    if valid == True:
        print(type)

    #prints if not valid
    else:
        print("INVALID")

    return 0



def credit_type(n):

    #set the first two numbers of the
    n = n[0] + n[1]

    #if first number is 4 return VISA
    if n[0] == '4':
        return "VISA"

    #if first two number is 51, 52, 53, 54, 55 return MASTER
    elif n == "51" or n == "52" or n == "53" or n == "54" or n == "55":
        return "MASTERCARD"

    #if first two number is 34 or 37 return AMEX
    elif n == "34" or n == "37":
        return "AMEX"

    #if can't identify return false
    else:
        return False

def check_sum(number, length):

    #cast the credit card number string to a int
    intNumber = int(number)

    #declare the sum variable
    sum = 0

    #declare the index that will be use to go through the number string
    index = length - 1

    #flag to determine whether it the every other number
    flag = 0

    #loops through the number string
    for i in range(length):

        #if every other number
        if flag == 1:

            #cast the number in question to an int
            temp = int(number[index])

            #multiply the number by 2
            temp *= 2

            #add to the sum variable
            sum += 2

            #truncate the number
            intNumber //= 10

            #reduce the index
            index -= 1

            #not it's is not the other number
            flag = 0

        #if the other number
        else:

            #truncate the last digit
            intNumber //= 10

            #move index back
            index -= 1

            #now is the other number
            flag = 1

    #return True if valid
    if sum % 2 == 0:
        return True

    #return False if not vaid
    else:
        return False




if __name__ == "__main__":
    main()

File: CS502018/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>

//Proto for whatCard function
void whatCard(long long num, int length);

int main(void)
{
    //declare credit number variable
    long long number = 0;

    //ask user for credit card number
    number = get_long_long("Enter the credit card number: ");

    //create a copy of a number to count the number
    long long creditCopy = number;
    int creditLength = 0;

    while(creditCopy != 0)
    {
        creditCopy /= 10;
        creditLength++;
    }

    //checks if the number is the proper length
    if(creditLength != 15 && creditLength != 16 && creditLength != 13)
    {
        printf("INVALID\n");
    }

    //flag for everyother number
    bool flag = false;

    //sum variable declared
    int sum = 0;

    //reset variable to credit card number
    creditCopy = number;

    //goes through each number of the credit number
    for(int i = 0; i < creditLength; i++)
    {

        //flag to check if it is every other number
        if(flag == true)
        {
            //gets the digit and times by 2
            int firstNum = ((creditCopy % 10) * 2);

            //if the number is greater than 10
            if(firstNum >= 10)
            {
                //goes through each digit and adds them
                for(int j = 0; j < 2; j++)
                {
                    sum = sum + (firstNum % 10);
                    firstNum /= 10;
                }
            }

            //adds number to sum
            else
            {
                sum = sum + firstNum;
            }

            //move on the next digit and make the flag false to skip the next one
            creditCopy /= 10;
            flag = false;
        }

        //get the digit and add it to sum
        else
        {
            int num = creditCopy % 10;
            sum = sum + num;
            creditCopy /= 10;
            flag = true;
        }
    }

    //checks if sum ends in 0
    if((sum % 10) == 0)
    {
        //call function to find out what card type it is
        whatCard(number, creditLength);
    }

    //if sum does not have 0 at the end print invalid
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

//whatCard prints what kind of card it is
void whatCard(long long num, int length)
{
    //declare array to store num in
    int numArray[length];

    //counter is length - 1 since arrays start at 0
    int counter = length - 1;

    //goes through each array and puts a variable from num
    for(int i = 0; i < length; i++)
    {
        numArray[counter] = (num % 10);
        counter--;
        num /= 10;
    }

    //check if card is a VISA
    if(numArray[0] == 4)
    {
        printf("VISA\n");
    }

    //checks if card is an American Express
    else if(numArray[0] == 3)
    {
        if(numArray[1] == 4 || numArray[1] == 7)
        {
            printf("AMEX\n");
        }

        else
        {
            printf("INVALID\n");
        }
    }

    //checks if card is a MasterCard
    else if(numArray[0] == 5)
    {
        if(numArray[1] == 1 ||numArray[1] == 2 ||numArray[1] == 3 ||numArray[1] == 4 ||numArray[1] == 5)
        {
            printf("MASTERCARD\n");
        }

        else
        {
            printf("INVALID\n");
        }
    }

    //prints invalid if no card starts with certain numbers
    else
    {
        printf("INVALID\n");
    }
}